src/app/util/attribute-storage.cpp

```diff
@@ -365,33 +365,46 @@ static uint8_t * singletonAttributeLocation(EmberAfAttributeMetadata * am)
 // This function does mem copy, but smartly, which means that if the type is a
 // string, it will copy as much as it can.
 // If src == NULL, then this method will set memory to zeroes
+// See documentation for emAfReadOrWriteAttribute for the semantics of
+// readLength when reading and writing.
 static EmberAfStatus typeSensitiveMemCopy(uint8_t * dest, uint8_t * src, EmberAfAttributeMetadata * am, bool write,
                                           uint16_t readLength)
 {
     EmberAfAttributeType attributeType = am->attributeType;
-    uint16_t size                      = (readLength == 0) ? am->size : readLength;
+    // readLength == 0 for a read indicates that we should just trust that the
+    // caller has enough space for an attribute...
+    bool ignoreReadLength = write || (readLength == 0);
```

Comment:
why is the caller buffer size ignored when write is set?

Generally I have complains about this method to start with: a buffer length that seems to be only relevant on read (called readLengt) but method used on both read and write and a direction parameter when the caller could just have flipped src and destination.

Could we fix up this method and make it somehow easier to use?

Reply:
> why is the caller buffer size ignored when write is set?

Apart from that being the documented behavior of `emAfReadOrWriteAttribute` and all callers who set `write` passing 0 for the size?  I'm afraid I don't know why the setup is the way it is; that's a question for @selissia or @tecimovic.

> Could we fix up this method and make it somehow easier to use?

Yes, we could, and I agree we should.  We'd need to audit all consumers to see what they are doing with `readLength`; the comments suggest that there are legacy consumers that pass 0 for reads as well....

I'm happy to do that, but it might be a few days before I can pull enough time off of spec stuff to do it.  I'm happy to land this in the meantime to unblock your ASan changes and do the larger semantic-changing refactoring in a followup, or expand the scope of this PR, as long as you're OK with the corresponding lag.  Please let me know.

Reply:
created #4394 to track improvement in buffer handling within this code.

Is this generated code? if so, how would one go about on changing the code and updating?

Reply:
Just to close the loop on this: this is not generated code.